src/xdsl/irdl.py

```diff
@@ -470,166 +486,237 @@ def from_pyrdl(pyrdl_def: type[_OpT]) -> OpDef:
         return op_def
 
 
+class ArgType(Enum):
+    """
+    An enum representing the part of an IR that may be variadic.
+    This contains operands, results, and regions.
+    """
+    OPERAND = 1
+    RESULT = 2
+    REGION = 3
+
+    def get_defs(
+        self, op_def: OpDef
+    ) -> list[tuple[str, OperandDef]] | list[tuple[str, ResultDef]] | list[
+            tuple[str, RegionDef]]:
+        """Get the definitions of this type in an operation definition."""
+        if self == self.OPERAND:
```

Comment:
This code looks extremely weird, but I think I get what you mean. Still, I wonder if this couldn't be implemented cleaner. Why should these methods be methods on the Enum?

Reply:
Yeah, I made these functions rather than methods.
The reason why I have these in the first place is to remove a lot of copy-paste (which is the root of most bugs).
I don't really know how to rewrite this in a better way (especially since there are no associated types in Python), but I'm extremely open for suggestions here!

Reply:
I prefer them as top-level functions. I think you missed the `get_name` one, right?

I don't really see a better approach either. Maybe doing something with a class hierarchy s.t. these checks become isinstance checks or fields of a class might be better. Something like ArgType is the parent and then there is OperandType, ResultType and RegionType with fields for name, arg_def, op_args and attr_size.

Reply:
I have a suggestion. Take ```get_name``` as an example.
Let ```WhatSoEverDef``` be a base class which has abstract function ```get_name``` and ```OperationDef``` etc inherit from it and implement the corresponding functions. Then any operation def inheriting ```OperationDef``` will automatically have these method implemented. Similar things can be done for ```get_arg_defs``` and ```get_op_args```

I don't see why we need associated type here yet.

Reply:
I kept `get_name` on purpose, since this is the only one that I feel make sense to keep in the enum. However, if you prefer it to be outside, that's fine for me too, since I don't have strong opinions against it!

Hmm, I kind of feel that using inheritance here would be going a bit too far in OOP. Mostly because none of these classes have any fields. If you both feel strongly that this would be better with inheritance, I'm okay in changing it. Otherwise, I would prefer keeping the enum, since I feel this fits better what we want (i.e. representing cases, rather than inheritance).

Reply:
Okay, I see why you kept it in. I still prefer having them outside like `map_x_to_y` should not be a method of `x`, but I don't have a strong opinion.

I feel like these classes would have to be renamed to actually nicely apply OOP. Maybe have `IRConstructs` and `VariadicIRConstructs` and then make Operand, Result and Region inherit from them?

I am also fine with keeping the enum, if you prefer it. I think you are right in that it represents cases in a more straightforward way.

Reply:
Good name suggestion!